File: benchmarks/swift/Sources/SwiftBenchmark/BenchmarkModels.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Fory
import Foundation

enum DataKind: String, CaseIterable {
    case numericStruct = "struct"
    case sample = "sample"
    case mediaContent = "mediacontent"
    case numericStructList = "structlist"
    case sampleList = "samplelist"
    case mediaContentList = "mediacontentlist"

    var title: String {
        switch self {
        case .numericStruct:
            return "NumericStruct"
        case .sample:
            return "Sample"
        case .mediaContent:
            return "MediaContent"
        case .numericStructList:
            return "NumericStructList"
        case .sampleList:
            return "SampleList"
        case .mediaContentList:
            return "MediaContentList"
        }
    }
}

enum SerializerKind: String, CaseIterable {
    case fory
    case protobuf
    case msgpack

    var title: String {
        rawValue.capitalized
    }
}

enum OperationKind: String, CaseIterable {
    case serialize
    case deserialize

    var title: String {
        rawValue.capitalized
    }
}

@ForyStruct
struct NumericStruct: Codable, Equatable {
    @ForyField(id: 1)
    var f1: Int32 = 0
    @ForyField(id: 2)
    var f2: Int32 = 0
    @ForyField(id: 3)
    var f3: Int32 = 0
    @ForyField(id: 4)
    var f4: Int32 = 0
    @ForyField(id: 5)
    var f5: Int32 = 0
    @ForyField(id: 6)
    var f6: Int32 = 0
    @ForyField(id: 7)
    var f7: Int32 = 0
    @ForyField(id: 8)
    var f8: Int32 = 0
    @ForyField(id: 9)
    var f9: Int32 = 0
    @ForyField(id: 10)
    var f10: Int32 = 0
    @ForyField(id: 11)
    var f11: Int32 = 0
    @ForyField(id: 12)
    var f12: Int32 = 0
}

@ForyStruct
struct Sample: Codable, Equatable {
    @ForyField(id: 1)
    var intValue: Int32 = 0
    @ForyField(id: 2)
    var longValue: Int64 = 0
    @ForyField(id: 3)
    var floatValue: Float = 0
    @ForyField(id: 4)
    var doubleValue: Double = 0
    @ForyField(id: 5)
    var shortValue: Int32 = 0
    @ForyField(id: 6)
    var charValue: Int32 = 0
    @ForyField(id: 7)
    var booleanValue: Bool = false
    @ForyField(id: 8)
    var intValueBoxed: Int32 = 0
    @ForyField(id: 9)
    var longValueBoxed: Int64 = 0
    @ForyField(id: 10)
    var floatValueBoxed: Float = 0
    @ForyField(id: 11)
    var doubleValueBoxed: Double = 0
    @ForyField(id: 12)
    var shortValueBoxed: Int32 = 0
    @ForyField(id: 13)
    var charValueBoxed: Int32 = 0
    @ForyField(id: 14)
    var booleanValueBoxed: Bool = false
    @ForyField(id: 15, type: .array(element: .int32()))
    var intArray: [Int32] = []
    @ForyField(id: 16, type: .array(element: .int64()))
    var longArray: [Int64] = []
    @ForyField(id: 17, type: .array(element: .float32))
    var floatArray: [Float] = []
    @ForyField(id: 18, type: .array(element: .float64))
    var doubleArray: [Double] = []
    @ForyField(id: 19, type: .array(element: .int32()))
    var shortArray: [Int32] = []
    @ForyField(id: 20, type: .array(element: .int32()))
    var charArray: [Int32] = []
    @ForyField(id: 21, type: .array(element: .bool))
    var booleanArray: [Bool] = []
    @ForyField(id: 22)
    var string: String = ""
}

@ForyEnum
enum Player: Int32, Codable, Equatable {
    case java = 0
    case flash = 1
}

@ForyEnum
enum Size: Int32, Codable, Equatable {
    case small = 0
    case large = 1
}

@ForyStruct
struct Media: Codable, Equatable {
    @ForyField(id: 1)
    var uri: String = ""
    @ForyField(id: 2)
    var title: String = ""
    @ForyField(id: 3)
    var width: Int32 = 0
    @ForyField(id: 4)
    var height: Int32 = 0
    @ForyField(id: 5)
    var format: String = ""
    @ForyField(id: 6)
    var duration: Int64 = 0
    @ForyField(id: 7)
    var size: Int64 = 0
    @ForyField(id: 8)
    var bitrate: Int32 = 0
    @ForyField(id: 9)
    var hasBitrate: Bool = false
    @ForyField(id: 10)
    var persons: [String] = []
    @ForyField(id: 11)
    var player: Player = .java
    @ForyField(id: 12)
    var copyright: String = ""
}

@ForyStruct
struct Image: Codable, Equatable {
    @ForyField(id: 1)
    var uri: String = ""
    @ForyField(id: 2)
    var title: String = ""
    @ForyField(id: 3)
    var width: Int32 = 0
    @ForyField(id: 4)
    var height: Int32 = 0
    @ForyField(id: 5)
    var size: Size = .small
}

@ForyStruct
struct MediaContent: Codable, Equatable {
    @ForyField(id: 1)
    var media: Media = .init()
    @ForyField(id: 2)
    var images: [Image] = []
}

@ForyStruct
struct NumericStructList: Codable, Equatable {
    @ForyField(id: 1)
    var structList: [NumericStruct] = []
}

@ForyStruct
struct SampleList: Codable, Equatable {
    @ForyField(id: 1)
    var sampleList: [Sample] = []
}

@ForyStruct
struct MediaContentList: Codable, Equatable {
    @ForyField(id: 1)
    var mediaContentList: [MediaContent] = []
}

enum BenchmarkDataFactory {
    static let listSize = 5

    static func makeNumericStruct() -> NumericStruct {
        NumericStruct(
            f1: -12345,
            f2: 987_654_321,
            f3: -31_415,
            f4: 27_182_818,
            f5: -32_000,
            f6: 1_000_000,
            f7: -999_999_999,
            f8: 42,
            f9: 123_456_789,
            f10: -42,
            f11: 31_415_926,
            f12: -27_182_818
        )
    }

    static func makeSample() -> Sample {
        Sample(
            intValue: 123,
            longValue: 1_230_000,
            floatValue: 12.345,
            doubleValue: 1.234_567,
            shortValue: 12_345,
            charValue: 33,
            booleanValue: true,
            intValueBoxed: 321,
            longValueBoxed: 3_210_000,
            floatValueBoxed: 54.321,
            doubleValueBoxed: 7.654_321,
            shortValueBoxed: 32_100,
            charValueBoxed: 36,
            booleanValueBoxed: false,
            intArray: [-1234, -123, -12, -1, 0, 1, 12, 123, 1234],
            longArray: [-123_400, -12_300, -1200, -100, 0, 100, 1200, 12_300, 123_400],
            floatArray: [-12.34, -12.3, -12.0, -1.0, 0.0, 1.0, 12.0, 12.3, 12.34],
            doubleArray: [-1.234, -1.23, -12.0, -1.0, 0.0, 1.0, 12.0, 1.23, 1.234],
            shortArray: [-1234, -123, -12, -1, 0, 1, 12, 123, 1234],
            charArray: [97, 115, 100, 102, 65, 83, 68, 70],
            booleanArray: [true, false, false, true],
            string: "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        )
    }

    static func makeMediaContent() -> MediaContent {
        MediaContent(
            media: Media(
                uri: "http://javaone.com/keynote.ogg",
                title: "",
                width: 641,
                height: 481,
                format: "video/theora\u{1234}",
                duration: 18_000_001,
                size: 58_982_401,
                bitrate: 0,
                hasBitrate: false,
                persons: ["Bill Gates, Jr.", "Steven Jobs"],
                player: .flash,
                copyright: "Copyright (c) 2009, Scooby Dooby Doo"
            ),
            images: [
                Image(
                    uri: "http://javaone.com/keynote_huge.jpg",
                    title: "Javaone Keynote\u{1234}",
                    width: 32_000,
                    height: 24_000,
                    size: .large
                ),
                Image(
                    uri: "http://javaone.com/keynote_large.jpg",
                    title: "",
                    width: 1024,
                    height: 768,
                    size: .large
                ),
                Image(
                    uri: "http://javaone.com/keynote_small.jpg",
                    title: "",
                    width: 320,
                    height: 240,
                    size: .small
                ),
            ]
        )
    }

    static func makeNumericStructList() -> NumericStructList {
        NumericStructList(structList: Array(repeating: makeNumericStruct(), count: listSize))
    }

    static func makeSampleList() -> SampleList {
        SampleList(sampleList: Array(repeating: makeSample(), count: listSize))
    }

    static func makeMediaContentList() -> MediaContentList {
        MediaContentList(mediaContentList: Array(repeating: makeMediaContent(), count: listSize))
    }
}

File: benchmarks/csharp/BenchmarkModels.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using Apache.Fory;
using MessagePack;
using ProtoBuf;

namespace Apache.Fory.Benchmarks.CSharp;

[ForyStruct]
[MessagePackObject(keyAsPropertyName: true)]
[ProtoContract]
public sealed class NumericStruct
{
    [ForyField(Id = 1)]
    [ProtoMember(1)]
    public int F1 { get; set; }

    [ForyField(Id = 2)]
    [ProtoMember(2)]
    public int F2 { get; set; }

    [ForyField(Id = 3)]
    [ProtoMember(3)]
    public int F3 { get; set; }

    [ForyField(Id = 4)]
    [ProtoMember(4)]
    public int F4 { get; set; }

    [ForyField(Id = 5)]
    [ProtoMember(5)]
    public int F5 { get; set; }

    [ForyField(Id = 6)]
    [ProtoMember(6)]
    public int F6 { get; set; }

    [ForyField(Id = 7)]
    [ProtoMember(7)]
    public int F7 { get; set; }

    [ForyField(Id = 8)]
    [ProtoMember(8)]
    public int F8 { get; set; }

    [ForyField(Id = 9)]
    [ProtoMember(9)]
    public int F9 { get; set; }

    [ForyField(Id = 10)]
    [ProtoMember(10)]
    public int F10 { get; set; }

    [ForyField(Id = 11)]
    [ProtoMember(11)]
    public int F11 { get; set; }

    [ForyField(Id = 12)]
    [ProtoMember(12)]
    public int F12 { get; set; }
}

[ForyStruct]
[MessagePackObject(keyAsPropertyName: true)]
[ProtoContract]
public sealed class NumericStructList
{
    [ForyField(Id = 1)]
    [ProtoMember(1)]
    public List<NumericStruct> Values { get; set; } = [];
}

[ForyStruct]
[MessagePackObject(keyAsPropertyName: true)]
[ProtoContract]
public sealed class Sample
{
    [ForyField(Id = 1)]
    [ProtoMember(1)]
    public int IntValue { get; set; }

    [ForyField(Id = 2)]
    [ProtoMember(2)]
    public long LongValue { get; set; }

    [ForyField(Id = 3)]
    [ProtoMember(3)]
    public float FloatValue { get; set; }

    [ForyField(Id = 4)]
    [ProtoMember(4)]
    public double DoubleValue { get; set; }

    [ForyField(Id = 5)]
    [ProtoMember(5)]
    public int ShortValue { get; set; }

    [ForyField(Id = 6)]
    [ProtoMember(6)]
    public int CharValue { get; set; }

    [ForyField(Id = 7)]
    [ProtoMember(7)]
    public bool BooleanValue { get; set; }

    [ForyField(Id = 8)]
    [ProtoMember(8)]
    public int IntValueBoxed { get; set; }

    [ForyField(Id = 9)]
    [ProtoMember(9)]
    public long LongValueBoxed { get; set; }

    [ForyField(Id = 10)]
    [ProtoMember(10)]
    public float FloatValueBoxed { get; set; }

    [ForyField(Id = 11)]
    [ProtoMember(11)]
    public double DoubleValueBoxed { get; set; }

    [ForyField(Id = 12)]
    [ProtoMember(12)]
    public int ShortValueBoxed { get; set; }

    [ForyField(Id = 13)]
    [ProtoMember(13)]
    public int CharValueBoxed { get; set; }

    [ForyField(Id = 14)]
    [ProtoMember(14)]
    public bool BooleanValueBoxed { get; set; }

    [ForyField(Id = 15)]
    [ProtoMember(15)]
    public int[] IntArray { get; set; } = [];

    [ForyField(Id = 16)]
    [ProtoMember(16)]
    public long[] LongArray { get; set; } = [];

    [ForyField(Id = 17)]
    [ProtoMember(17)]
    public float[] FloatArray { get; set; } = [];

    [ForyField(Id = 18)]
    [ProtoMember(18)]
    public double[] DoubleArray { get; set; } = [];

    [ForyField(Id = 19)]
    [ProtoMember(19)]
    public int[] ShortArray { get; set; } = [];

    [ForyField(Id = 20)]
    [ProtoMember(20)]
    public int[] CharArray { get; set; } = [];

    [ForyField(Id = 21)]
    [ProtoMember(21)]
    public bool[] BooleanArray { get; set; } = [];

    [ForyField(Id = 22)]
    [ProtoMember(22)]
    public string String { get; set; } = string.Empty;
}

[ForyStruct]
[MessagePackObject(keyAsPropertyName: true)]
[ProtoContract]
public sealed class SampleList
{
    [ForyField(Id = 1)]
    [ProtoMember(1)]
    public List<Sample> Values { get; set; } = [];
}

[ForyStruct]
[ProtoContract]
public enum Player
{
    [ProtoEnum]
    Java,
    [ProtoEnum]
    Flash,
}

[ForyStruct]
[ProtoContract]
public enum MediaSize
{
    [ProtoEnum]
    Small,
    [ProtoEnum]
    Large,
}

[ForyStruct]
[MessagePackObject(keyAsPropertyName: true)]
[ProtoContract]
public sealed class Media
{
    [ForyField(Id = 1)]
    [ProtoMember(1)]
    public string Uri { get; set; } = string.Empty;

    [ForyField(Id = 2)]
    [ProtoMember(2)]
    public string Title { get; set; } = string.Empty;

    [ForyField(Id = 3)]
    [ProtoMember(3)]
    public int Width { get; set; }

    [ForyField(Id = 4)]
    [ProtoMember(4)]
    public int Height { get; set; }

    [ForyField(Id = 5)]
    [ProtoMember(5)]
    public string Format { get; set; } = string.Empty;

    [ForyField(Id = 6)]
    [ProtoMember(6)]
    public long Duration { get; set; }

    [ForyField(Id = 7)]
    [ProtoMember(7)]
    public long Size { get; set; }

    [ForyField(Id = 8)]
    [ProtoMember(8)]
    public int Bitrate { get; set; }

    [ForyField(Id = 9)]
    [ProtoMember(9)]
    public bool HasBitrate { get; set; }

    [ForyField(Id = 10)]
    [ProtoMember(10)]
    public List<string> Persons { get; set; } = [];

    [ForyField(Id = 11)]
    [ProtoMember(11)]
    public Player Player { get; set; }

    [ForyField(Id = 12)]
    [ProtoMember(12)]
    public string Copyright { get; set; } = string.Empty;
}

[ForyStruct]
[MessagePackObject(keyAsPropertyName: true)]
[ProtoContract]
public sealed class Image
{
    [ForyField(Id = 1)]
    [ProtoMember(1)]
    public string Uri { get; set; } = string.Empty;

    [ForyField(Id = 2)]
    [ProtoMember(2)]
    public string Title { get; set; } = string.Empty;

    [ForyField(Id = 3)]
    [ProtoMember(3)]
    public int Width { get; set; }

    [ForyField(Id = 4)]
    [ProtoMember(4)]
    public int Height { get; set; }

    [ForyField(Id = 5)]
    [ProtoMember(5)]
    public MediaSize Size { get; set; }
}

[ForyStruct]
[MessagePackObject(keyAsPropertyName: true)]
[ProtoContract]
public sealed class MediaContent
{
    [ForyField(Id = 1)]
    [ProtoMember(1)]
    public Media Media { get; set; } = new();

    [ForyField(Id = 2)]
    [ProtoMember(2)]
    public List<Image> Images { get; set; } = [];
}

[ForyStruct]
[MessagePackObject(keyAsPropertyName: true)]
[ProtoContract]
public sealed class MediaContentList
{
    [ForyField(Id = 1)]
    [ProtoMember(1)]
    public List<MediaContent> Values { get; set; } = [];
}

public static class BenchmarkDataFactory
{
    private const int ListSize = 5;

    public static NumericStruct CreateNumericStruct()
    {
        return new NumericStruct
        {
            F1 = -12345,
            F2 = 987654321,
            F3 = -31415,
            F4 = 27182818,
            F5 = -32000,
            F6 = 1000000,
            F7 = -999999999,
            F8 = 42,
            F9 = 123456789,
            F10 = -42,
            F11 = 31415926,
            F12 = -27182818,
        };
    }

    public static Sample CreateSample()
    {
        return new Sample
        {
            IntValue = 123,
            LongValue = 1230000,
            FloatValue = 12.345f,
            DoubleValue = 1.234567,
            ShortValue = 12345,
            CharValue = '!',
            BooleanValue = true,
            IntValueBoxed = 321,
            LongValueBoxed = 3210000,
            FloatValueBoxed = 54.321f,
            DoubleValueBoxed = 7.654321,
            ShortValueBoxed = 32100,
            CharValueBoxed = '$',
            BooleanValueBoxed = false,
            IntArray = [-1234, -123, -12, -1, 0, 1, 12, 123, 1234],
            LongArray = [-123400, -12300, -1200, -100, 0, 100, 1200, 12300, 123400],
            FloatArray = [-12.34f, -12.3f, -12.0f, -1.0f, 0.0f, 1.0f, 12.0f, 12.3f, 12.34f],
            DoubleArray = [-1.234, -1.23, -12.0, -1.0, 0.0, 1.0, 12.0, 1.23, 1.234],
            ShortArray = [-1234, -123, -12, -1, 0, 1, 12, 123, 1234],
            CharArray = ['a', 's', 'd', 'f', 'A', 'S', 'D', 'F'],
            BooleanArray = [true, false, false, true],
            String = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789",
        };
    }

    public static MediaContent CreateMediaContent()
    {
        return new MediaContent
        {
            Media = new Media
            {
                Uri = "http://javaone.com/keynote.ogg",
                Title = string.Empty,
                Width = 641,
                Height = 481,
                Format = "video/theora\u1234",
                Duration = 18_000_001,
                Size = 58_982_401,
                Bitrate = 0,
                HasBitrate = false,
                Persons = ["Bill Gates, Jr.", "Steven Jobs"],
                Player = Player.Flash,
                Copyright = "Copyright (c) 2009, Scooby Dooby Doo",
            },
            Images =
            [
                new Image
                {
                    Uri = "http://javaone.com/keynote_huge.jpg",
                    Title = "Javaone Keynote\u1234",
                    Width = 32_000,
                    Height = 24_000,
                    Size = MediaSize.Large,
                },
                new Image
                {
                    Uri = "http://javaone.com/keynote_large.jpg",
                    Title = string.Empty,
                    Width = 1_024,
                    Height = 768,
                    Size = MediaSize.Large,
                },
                new Image
                {
                    Uri = "http://javaone.com/keynote_small.jpg",
                    Title = string.Empty,
                    Width = 320,
                    Height = 240,
                    Size = MediaSize.Small,
                },
            ],
        };
    }

    public static NumericStructList CreateNumericStructList()
    {
        NumericStructList list = new();
        for (int i = 0; i < ListSize; i++)
        {
            list.Values.Add(CreateNumericStruct());
        }

        return list;
    }

    public static SampleList CreateSampleList()
    {
        SampleList list = new();
        for (int i = 0; i < ListSize; i++)
        {
            list.Values.Add(CreateSample());
        }

        return list;
    }

    public static MediaContentList CreateMediaContentList()
    {
        MediaContentList list = new();
        for (int i = 0; i < ListSize; i++)
        {
            list.Values.Add(CreateMediaContent());
        }

        return list;
    }
}
